### DISCUSSION TITLE
Winston + seq custom log levels?

### DISCUSSION THREAD

**Discussion by opadfnezig** [datalust/seq-tickets#2098] [Help] [answered]
Using seq with winston(winston-seq) in my node application. Trying to configure custom log levels and having issues with that. It just doesn't work, and I wasn't able to find a solution that would. Tried different solutions, like example config from winston documentation, different libraries..

Though, I can create filters for thouse custom levels in the web interface, so it should be possible I guess?

There are only 3 levels that seems to be working: error, warn, info. I would like to have at least http and database, or verbose level for both.

Can't find any information about log levels in updated documentation. 

Need any help/information about how to create log levels or alternatives.

**Comment by nblumhardt** [comment]
Hi @opadfnezig, thanks for reaching out.

Can you please post a snippet showing what you're trying, and we can take a look? There's nothing in the `winston-seq` package that we're aware of that will have any impact on the allowable levels.

Cheers,
Nick

**Comment by opadfnezig** [comment]
Sure.

*Example 1, trying to use winston's  example code with winston-seq*
```
const winston = require('winston');
const { SeqTransport } = require('@datalust/winston-seq');

const myCustomLevels = {
    levels: {
        foo: 0,
        bar: 1,
        baz: 2,
        foobar: 3
    },
    colors: {
        foo: 'blue',
        bar: 'green',
        baz: 'yellow',
        foobar: 'red'
    }
};

const customLevelLogger = winston.createLogger({
    defaultMeta: {
        service: 'tester'
    },
    transports: [
        new SeqTransport({
            format: winston.format.simple(),
            serverUrl: 'http://127.0.0.1:5341',
            apiKey: '12345678901234567890',
        })
    ],
    levels: myCustomLevels.levels
});

customLevelLogger.log('foo', 'foo');
customLevelLogger.log('bar', 'bar');
customLevelLogger.log('baz', 'baz');
customLevelLogger.log('foobar', 'foobar');
```

*Example 2. Trying default levels.*

```
const winston = require('winston');
const { SeqTransport } = require('@datalust/winston-seq');
/*
const myCustomLevels = {
    levels: {
        foo: 0,
        bar: 1,
        baz: 2,
        foobar: 3
    },
    colors: {
        foo: 'blue',
        bar: 'green',
        baz: 'yellow',
        foobar: 'red'
    }
};
*/
const winston = require('winston');
const { SeqTransport } = require('@datalust/winston-seq');

const logger = winston.createLogger({
    /*format: winston.format.combine(
        winston.format.colorize(),
        winston.format.errors({ stack: true }),
        winston.format.json(),
    ),*/
    defaultMeta: {
        service: 'tester'
    },
    transports: [
        new SeqTransport({
            format: winston.format.simple(),
            serverUrl: 'http://127.0.0.1:5341',
            apiKey: '12345678901234567890',
            //onError: (e => { console.error(e) }),
        })
    ],
});

logger.error('test error'); //+
logger.warn('test warn'); //+
logger.info('test info'); //+
logger.log('debug', 'test debug');
logger.debug('test debug');
logger.log('silly', 'test silly');
logger.silly('test silly');
logger.log('http', 'test http');

```

What I receive from running all the code (both examples)
![image](https://github.com/datalust/seq-tickets/assets/38683305/68161ee2-6d58-452c-b2f0-ba82650e3d32)

**Reply by liammclennan** [reply]
If you use a different transport (console, file etc) do you see all the logs?

Do you have any signals active in Seq?

**Reply by opadfnezig** [reply]
Tried out example code from winston.
```
const winston = require('winston');
const { SeqTransport } = require('@datalust/winston-seq');

const myCustomLevels = {
    levels: {
        foo: 0,
        bar: 1,
        baz: 2,
        foobar: 3
    },
    colors: {
        foo: 'blue',
        bar: 'green',
        baz: 'yellow',
        foobar: 'red'
    }
};

const customLevelLogger = winston.createLogger({
    levels: myCustomLevels.levels
});

customLevelLogger.foobar('some foobar level-ed message');
```

This works. But if I add seq transport - no.

Just checked out this in default console winston
```
logger.error('test error'); //+
logger.warn('test warn'); //+
logger.info('test info'); //+
logger.log('debug', 'test debug');
logger.debug('test debug');
logger.log('silly', 'test silly');
logger.silly('test silly');
logger.log('http', 'test http');
```

Works same as with seq, only first three does something.

I've added signal for http, but not sure if I've done it right. And it doesn't change anything.
![image](https://github.com/datalust/seq-tickets/assets/38683305/f5b01ea9-7571-478b-b88f-96454b53527e)

**Comment by liammclennan** [comment, marked as answer]
Hi @opadfnezig,

I fixed your example and it logs at custom levels:

```js
const winston = require('winston');
const { SeqTransport } = require('@datalust/winston-seq');

const myCustomLevels = {
    levels: {
        foo: 0,
        bar: 1,
        baz: 2,
        foobar: 3
    },
    colors: {
        foo: 'blue',
        bar: 'green',
        baz: 'yellow',
        foobar: 'red'
    }
};

const customLevelLogger = winston.createLogger({
    level: 'foobar',
    format: winston.format.simple(),
    transports: [
        new SeqTransport({
            serverUrl: 'http://localhost:5341',
        })
    ],
    levels: myCustomLevels.levels
});

customLevelLogger.log('foo', 'foo');
customLevelLogger.log('bar', 'bar');
customLevelLogger.log('baz', 'baz');
customLevelLogger.log('foobar', 'foobar');
```

![image](https://github.com/datalust/seq-tickets/assets/271514/1c6ab3b5-1d84-4c18-8e91-3abb0f7778e5)

Hope this helps.

**Reply by opadfnezig** [reply]
Thank you. It works.

And I finally understand what is the problem. Probably I should create an issue.

What is that level for? Is that default level like, if you do` logger.log('a')` it will log to `foobar`?

Seems like if you don't specify the default level it woun't work.
My code works well If I add level property, and does nothing + executes too fast (not sure what's actually happening there) if there's no level property specified.

**Reply by liammclennan** [reply]
Winston is a separate library, not maintained by Datalust, so you might want to ask on [their repo](https://github.com/winstonjs/winston). From [their docs](https://github.com/winstonjs/winston?tab=readme-ov-file#creating-your-own-logger) it looks like the `level` property establishes the maximum log level that will get logged.
